spec/v1.0-draft/resource_descriptor.md

```diff
@@ -0,0 +1,167 @@
+# ResourceDescriptor field type specification
+
+Version v1.0-draft
+
+A size-efficient description of any software artifact or resource (mutable
+or immutable).
+
+## Schema
+
+```json
+{
+  "name": "<NAME>",
+  "uri": "<RESOURCE URI>",
+  "digest": { "<ALGORITHM>": "<HEX VALUE>", ... },
+  "content": "<BASE64-ENCODED BYTES>",
+  "downloadLocation": "<RESOURCE URI>"
+  "mediaType": "<MIME TYPE>",
+  "annotations": { /* object */ }
+}
+```
+
+## Fields
+
+Though all fields are optional, a ResourceDescriptor MUST specify one of
+`uri` or `digest` at a minimum.
+
+`name` _string, optional_
+
+> Machine-readable identifier for distinguishing between descriptors.
+>
+> The semantics are up to the producer and consumer. The `name` SHOULD be
+> stable, such as a filename, to allow consumers to reliably use the `name`
+> as part of their policy.
+
+`uri` _[ResourceURI], optional_
+
+> A URI used to identify the resource or artifact globally.
+>
+> When a `digest`cannot be computed for the resource or artifact, the
+> producer MUST set this field.
+
+`digest` _[DigestSet], optional_
+
+> A set of cryptographic digests of the contents of the resource or artifact.
+>
+> When known, the producer SHOULD set this field to denote an immutable
+> artifact or resource. The producer and consumer must agree on acceptable
+> algorithms.
+
+`content` _base64-encoded bytes, optional_
+
+> The contents of the resource or artifact.
+>
+> The producer MAY use this field in scenarios where including the contents
+> of the resource/artifact directly in the attestation is deemed more
+> efficient for consumers than providing a pointer to another location. To
+> maintain size efficiency, the size of `content` SHOULD be less than 1KB.
+>
+> The producer and consumer must agree on the semantics. The `uri` or
+> `mediaType` MAY be used by the producer as hints for how consumers should
+> parse `content`.
+
+`downloadLocation` _[ResourceURI], optional_
+
+> The location of the described resource or artifact, if different from the
+> `uri`.
+>
+> To enable automated downloads by consumers, the specified location SHOULD
+> be resolvable.
+
+`mediaType` _string, optional_
+
+> The [MIME Type][] (i.e., media type) of the described resource or artifact.
+>
+> For resources or artifacts that do not have a standardized MIME type,
+> producers SHOULD follow [RFC 6838] (Sections 3.2-3.4) conventions of
+> prefixing types with `x.`, `prs.`, or `vnd.` to avoid collisions with other
+> producers.
+
+`annotations` _object, optional_
```

Comment:
Oh maybe where I've been getting confused is I've assumed 'annotations' is a map<string, object> because I assumed folks might have multiple different annotations they might want to apply.  So instead of 

```jsonl
"annotations": { "_type": "x.git.review", "twoPartyReview": false }
```

I'd expect something like

```jsonl
"annotations": {
    "x.git.review": { "twoPartyReview": false },
    "vnd.acme.scan": { "fuzzed": false, "licenseCompliant": true }
}
```

If you agree we should probably strike the `_type` suggestion (since it would prevent multiple annotations).

The text might the be something like

>  The producer and consumer must agree on the semantics and acceptable
> formats for the `annotations` objects. There are currently no standardized
> annotation formats. To enable consumers to parse the field, `annotations`
> keys SHOULD be self-describing via a name following the conventions of
> custom MIME types per [RFC 6838] Sections 3.2-3.4 (e.g. `x.`, `prs.`, `vnd.`).

WDYT?

Reply:
a) Agree that the format should be map <string, object>, so will make those changes
b) Given #147, I'm almost wondering if we should avoid giving any kind of formatting guidelines for the annotations or extensions fields other than "avoid dots and dollar-signs in the strings" (field names with '$' are also not well-supported before MongoDB v5).